File: archive/B&B/InputData.cs
using Serilog;
using System.Text;
using static System.Environment;

namespace Bb4Ssrp
{
    public static class InputData
    {
        public static string[] words = { };
        public static string alphabet = "";
        public static int m = 0;  // the size of a word
        public static int n = 0;  // the number of words
        public static int k = 2;  // the number of columns to remove

        public static void readData(string fileName)
        {
            try
            {
                words = File.ReadAllLines(fileName);
            }
            catch (Exception ex)
            {
                Log.Error("There is a problem with file {A}:", fileName);
                Log.Error("{A} says {B}", ex.GetType(), ex.Message);
                Exit(1);
            }
            n = words.Length;
            if (n == 0)
            {
                Log.Error("There are no words in file {A}.", fileName);
                Exit(1);
            }
            m = words[0].Length;
            Log.Debug("The length of the first word is {A}.", m);
            foreach (string w in words)
            {
                if (w.Length != m)
                {
                    Log.Error("Not all the strings have the same length!");
                    Exit(1);
                }
                foreach (char symbol in w)
                {
                    if (!alphabet.Contains(symbol))
                    {
                        alphabet += symbol;
                    }
                }
            }

        }

        public static int[] determineWordIndices(int[] arrOfCols)
        {
            Dictionary<string, HashSet<int>> result = new();
            var columns = arrOfCols.ToHashSet();
            for (int i = 0; i < n; i++)
            {
                StringBuilder str = new();
                for (int j = 0; j < m; j++)
                {
                    if (!columns.Contains(j))
                    {
                        str.Append(words[i][j]);
                    }
                }
                var s = str.ToString();
                if (!result.ContainsKey(s))
                {
                    result.Add(s, new HashSet<int>());
                }
                result[s].Add(i);
            }
            var (w, inds) = result.MaxBy(entry => entry.Value.Count);
            return inds.ToArray();
        }

        public static int determineWordSetSize(HashSet<int> columns)
        {
            Dictionary<string, int> result = new();
            for (int i = 0; i < n; i++)
            {
                StringBuilder str = new();
                for (int j = 0; j < m; j++)
                {
                    if (!columns.Contains(j))
                    {
                        str.Append(words[i][j]);
                    }
                }
                var s = str.ToString();
                if (!result.ContainsKey(s))
                {
                    result.Add(s, 0);
                }
                ++result[s];
            }
            var (_, size) = result.MaxBy(entry => entry.Value);
            return size;
        }
    }
}

File: archive/UCT/src/main/scala/ssrp/InputData.scala
package ssrp

import com.typesafe.scalalogging._
import scala.collection.mutable
import scala.io.Source

object InputData extends StrictLogging {
  var words: Array[String] = null
  var m = 0  // the number of words
  var n = 0  // the size of a word

  def readData(fileName: String): Unit = {
    val source = Source.fromFile(fileName)
    words = try source.getLines.toArray finally source.close()
    if(words eq null) {
      logger.error(s"There is a problem with file $fileName.")
      sys.exit(1)
    }
    m = words.length
    if(m == 0) {
      logger.error(s"There are no words in file $fileName.")
      sys.exit(1)
    }
    n = words(0).length
    logger.debug(s"The length of the first word is $n.")
    for(w <- words) {
      if(w.length != n) {
        logger.error("Not all the strings have the same length!")
        sys.exit(1)
      }
    }
  }

  def determineWordIndices(arrOfCols: Array[Int]): Array[Int] = {
    val result = mutable.HashMap.empty[String, mutable.Set[Int]]
    val columns = mutable.HashSet.from(arrOfCols)
    for(i <- 0 until m) {
      val str = new mutable.StringBuilder()
      for(j <- 0 until n if !columns.contains(j)) {
        str.addOne(InputData.words(i)(j))
      }
      val s = str.toString
      if(result.contains(s))
        result(s).addOne(i)
      else
        result(s) = mutable.HashSet(i)
    }
    val (w, inds) = result.maxBy(entry => entry._2.size)
    inds.toArray
  }
}
